operator/cmd/k8s_identity.go

```diff
@@ -10,78 +10,208 @@ import (
+		IdentityGCInterval:       operatorCfg.IdentityGCInterval,
+		IdentityHeartbeatTimeout: operatorCfg.IdentityHeartbeatTimeout,
+		EndpointGCInterval:       operatorCfg.EndpointGCInterval,
+		RateInterval:             operatorCfg.IdentityGCRateInterval,
+		RateLimit:                operatorCfg.IdentityGCRateLimit,
+		MetricsEnabled:           operatorCfg.EnableMetrics,
 	}
-	err = clientset.CiliumV2().CiliumIdentities().Delete(
-		ctx,
-		identity.Name,
-		metav1.DeleteOptions{
-			Preconditions: &metav1.Preconditions{
-				UID:             &identity.UID,
-				ResourceVersion: &identity.ResourceVersion,
-			},
-		})
-	if err != nil {
-		log.WithError(err).Error("Unable to delete identity")
-	} else {
-		log.WithField(logfields.Identity, identity.GetName()).Info("Garbage collected identity")
+}
+
+// IdentityGCDeps contains all the dependencies for the identity-gc.
+// They will be provided through dependency injection.
+type IdentityGCDeps struct {
+	cell.In
+
+	Logger logrus.FieldLogger
+	LC     hive.Lifecycle
+
+	Clientset        k8sClient.Clientset
+	IdentityResource resource.Resource[*v2.CiliumIdentity]
+}
+
+func newIdentitiesGC(deps IdentityGCDeps, cfg IdentitiesGCConfig) *IdentitiesGC {
+	obj := &IdentitiesGC{
+		cfg:              cfg,
+		logger:           deps.Logger,
+		clientset:        deps.Clientset,
+		identityResource: deps.IdentityResource,
+		heartbeatStore: identity.NewIdentityHeartbeatStore(
+			cfg.IdentityHeartbeatTimeout,
+		),
+		rateLimiter: rate.NewLimiter(
+			cfg.RateInterval,
+			cfg.RateLimit,
+		),
 	}
 
-	return err
+	deps.LC.Append(obj)
+
+	return obj
 }
 
-func updateIdentity(ctx context.Context, clientset k8sClient.Clientset, identity *v2.CiliumIdentity) error {
-	_, err := clientset.CiliumV2().CiliumIdentities().Update(
-		ctx,
-		identity,
-		metav1.UpdateOptions{})
-	if err != nil {
-		log.WithError(err).Error("Updating Identity")
-	} else {
-		log.WithField(logfields.Identity, identity.GetName()).Debug("Updated identity")
+// IdentitiesGC represents the Cilium identities periodic GC.
+type IdentitiesGC struct {
+	logger logrus.FieldLogger
+
+	cfg IdentitiesGCConfig
+
+	clientset        k8sClient.Clientset
+	identityResource resource.Resource[*v2.CiliumIdentity]
+
+	heartbeatStore *identity.IdentityHeartbeatStore
+	mgr            *controller.Manager
+	rateLimiter    *rate.Limiter
+
+	// counters for GC failed/successful runs
+	failedRuns     float64
+	successfulRuns float64
+
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+// Start implements hive.HookInterface
+func (igc *IdentitiesGC) Start(ctx hive.HookContext) error {
+	if igc.cfg.AllocationMode != option.IdentityAllocationModeCRD {
+		return nil
+	}
+
+	if !igc.clientset.IsEnabled() {
+		igc.logger.Fatal("CRD Identity allocation mode requires k8s to be configured.")
+	}
+
+	if igc.cfg.EndpointGCInterval == 0 {
+		log.Fatal("The CiliumIdentity garbage collector requires the CiliumEndpoint garbage collector to be enabled")
```

Comment:
Hmm, any idea why this is necessary? I wonder if things like this (e.g. cross-module dependencies) should be solved outside, e.g. in `root.go`.

Reply:
Moved the check outside, in `operator/cmd/root.go`. TBH, I don't know why this is necessary, though. Needs to investigate.